File: backend/app/schemas/kreisverband.py
"""Kreisverband Pydantic schemas"""
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import date, datetime


class KreisverbandBase(BaseModel):
    name: str
    kuerzel: Optional[str] = None
    email: Optional[str] = None
    ist_aktiv: bool = True
    tenant_id: Optional[int] = None


class KreisverbandCreate(KreisverbandBase):
    pass


class KreisverbandUpdate(BaseModel):
    name: Optional[str] = None
    kuerzel: Optional[str] = None
    email: Optional[str] = None
    ist_aktiv: Optional[bool] = None
    tenant_id: Optional[int] = None


class KVVorstandsmitgliedBase(BaseModel):
    name: str
    email: Optional[str] = None
    rolle: str
    amtszeit_start: Optional[date] = None
    amtszeit_ende: Optional[date] = None
    ist_aktiv: bool = True


class KVVorstandsmitgliedCreate(KVVorstandsmitgliedBase):
    kreisverband_id: int


class KVVorstandsmitgliedResponse(KVVorstandsmitgliedBase):
    id: int
    kreisverband_id: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class KVProtokollBase(BaseModel):
    titel: str
    datum: date
    typ: str
    beschreibung: Optional[str] = None


class KVProtokollCreate(KVProtokollBase):
    kreisverband_id: int


class KVProtokollResponse(KVProtokollBase):
    id: int
    kreisverband_id: int
    datei_pfad: Optional[str] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class KreisverbandResponse(KreisverbandBase):
    id: int
    created_at: datetime
    updated_at: datetime
    vorstandsmitglieder: List[KVVorstandsmitgliedResponse] = []
    model_config = ConfigDict(from_attributes=True)


class KreisverbandListResponse(KreisverbandBase):
    id: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Vorstandsübersicht Landesverband: pro KV die gefilterten Vorstandsmitglieder
class VorstandUebersichtMitglied(BaseModel):
    id: int
    name: str
    email: Optional[str] = None
    rolle: str


class VorstandUebersichtEintrag(BaseModel):
    kreisverband: KreisverbandListResponse
    mitglieder: List[VorstandUebersichtMitglied]

File: frontend/src/lib/api/kreisverband.ts
import apiClient from './client';

export interface Kreisverband {
  id: number;
  name: string;
  kuerzel?: string | null;
  email?: string | null;
  ist_aktiv: boolean;
  tenant_id?: number | null;
  created_at: string;
  updated_at?: string;
  vorstandsmitglieder?: Vorstandsmitglied[];
}

export interface Vorstandsmitglied {
  id: number;
  kreisverband_id: number;
  name: string;
  email?: string | null;
  rolle: string;
  amtszeit_start?: string | null;
  amtszeit_ende?: string | null;
  ist_aktiv: boolean;
  created_at: string;
}

export interface KVProtokoll {
  id: number;
  kreisverband_id: number;
  titel: string;
  datum: string;
  typ: string;
  beschreibung?: string | null;
  datei_pfad?: string | null;
  created_at: string;
}

/** Rolle für Landesverband-Vorstandsübersicht */
export type VorstandUebersichtRolle = 'vorsitz' | 'schatzmeister' | 'organisation' | 'programmatik' | 'presse';

export interface VorstandUebersichtMitglied {
  id: number;
  name: string;
  email?: string | null;
  rolle: string;
}

export interface VorstandUebersichtEintrag {
  kreisverband: Pick<Kreisverband, 'id' | 'name' | 'kuerzel'> & { created_at?: string };
  mitglieder: VorstandUebersichtMitglied[];
}

export async function getLandesverbandVorstandUebersicht(params: {
  rolle: VorstandUebersichtRolle;
  mit_beisitzern?: boolean;
}): Promise<VorstandUebersichtEintrag[]> {
  const response = await apiClient.get<VorstandUebersichtEintrag[]>('/kreisverband/landesverband/vorstand-uebersicht', {
    params: { rolle: params.rolle, mit_beisitzern: params.mit_beisitzern ?? true },
  });
  return response.data;
}

export async function getKreisverbande(params?: { ist_aktiv?: boolean }): Promise<Kreisverband[]> {
  const response = await apiClient.get<Kreisverband[]>('/kreisverband/', { params });
  return response.data;
}

export async function getKreisverbandById(id: number): Promise<Kreisverband> {
  const response = await apiClient.get<Kreisverband>(`/kreisverband/${id}`);
  return response.data;
}

export async function createKreisverband(data: Partial<Kreisverband>): Promise<Kreisverband> {
  const response = await apiClient.post<Kreisverband>('/kreisverband/', data);
  return response.data;
}

export async function updateKreisverband(id: number, data: Partial<Kreisverband>): Promise<Kreisverband> {
  const response = await apiClient.put<Kreisverband>(`/kreisverband/${id}`, data);
  return response.data;
}

export async function deleteKreisverband(id: number): Promise<void> {
  await apiClient.delete(`/kreisverband/${id}`);
}

export async function getVorstand(kvId: number): Promise<Vorstandsmitglied[]> {
  const response = await apiClient.get<Vorstandsmitglied[]>(`/kreisverband/${kvId}/vorstand`);
  return response.data;
}

export async function createVorstandsmitglied(kvId: number, data: Omit<Vorstandsmitglied, 'id' | 'kreisverband_id' | 'created_at'>): Promise<Vorstandsmitglied> {
  const response = await apiClient.post<Vorstandsmitglied>(`/kreisverband/${kvId}/vorstand`, {
    ...data,
    kreisverband_id: kvId,
  });
  return response.data;
}

export async function updateVorstandsmitglied(mitgliedId: number, data: Omit<Vorstandsmitglied, 'id' | 'kreisverband_id' | 'created_at'>): Promise<Vorstandsmitglied> {
  const response = await apiClient.put<Vorstandsmitglied>(`/kreisverband/vorstand/${mitgliedId}`, data);
  return response.data;
}

export async function deleteVorstandsmitglied(mitgliedId: number): Promise<void> {
  await apiClient.delete(`/kreisverband/vorstand/${mitgliedId}`);
}

export async function getProtokolle(kvId: number, params?: { typ?: string }): Promise<KVProtokoll[]> {
  const response = await apiClient.get<KVProtokoll[]>(`/kreisverband/${kvId}/protokolle`, { params });
  return response.data;
}

export async function uploadProtokoll(kvId: number, formData: FormData): Promise<KVProtokoll> {
  const response = await apiClient.post<KVProtokoll>(`/kreisverband/${kvId}/protokolle`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data;
}

export async function deleteProtokoll(protokollId: number): Promise<void> {
  await apiClient.delete(`/kreisverband/protokolle/${protokollId}`);
}
